Bugzilla/ModPerl.pm

```diff
@@ -14,12 +14,101 @@ use File::Find ();
 use Cwd ();
 use Carp ();
 
-# We don't need (or want) to use Bugzilla's template subclass.
-# it is easier to reason with the code without all the extra things Bugzilla::Template adds
-# (and there might be side-effects, since this code is loaded very early in the httpd startup)
-use Template ();
-
 use Bugzilla::ModPerl::BlockIP;
+use Bugzilla::ModPerl::ResponseHandler;
+use Bugzilla::ModPerl::CleanupHandler;
+use Bugzilla::Constants qw(USE_NYTPROF);
+
+use Apache2::Log ();
+use Apache2::ServerUtil;
+use Apache2::SizeLimit;
+use ModPerl::RegistryLoader ();
+use File::Basename ();
+use File::Find ();
+
+use Bugzilla ();
+use Bugzilla::BugMail ();
+use Bugzilla::CGI ();
+use Bugzilla::Extension ();
+use Bugzilla::Install::Requirements ();
+use Bugzilla::Util ();
+use Bugzilla::RNG ();
+
+BEGIN {
+    if (USE_NYTPROF) {
+        $ENV{NYTPROF} = "savesrc=0:start=no:addpid=1";
+    }
+}
+use if USE_NYTPROF, 'Devel::NYTProf::Apache';
+
+# Make warnings go to the virtual host's log and not the main
+# server log.
+BEGIN { *CORE::GLOBAL::warn = \&Apache2::ServerRec::warn; }
+
+sub startup {
+
+    # Pre-compile the CGI.pm methods that we're going to use.
+    Bugzilla::CGI->compile(qw(:cgi :push));
+
+    # Pre-load localconfig. It might already be loaded, but we need to make sure.
+    Bugzilla->localconfig;
+    if ( $ENV{LOCALCONFIG_ENV} ) {
+        delete @ENV{ (Bugzilla::Install::Localconfig::ENV_KEYS) };
+    }
+
+    # This means that every httpd child will die after processing a request if it
+    # is taking up more than $apache_size_limit of RAM all by itself, not counting RAM it is
+    # sharing with the other httpd processes.
+    my $limit = Bugzilla->localconfig->{apache_size_limit};
+    if ( $limit < 400_000 ) {
+        $limit = 400_000;
+    }
+    Apache2::SizeLimit->set_max_unshared_size($limit);
+
+    my $cgi_path = Bugzilla::Constants::bz_locations()->{'cgi_path'};
+
+    # Set up the configuration for the web server
+    my $server = Apache2::ServerUtil->server;
+    my $conf   = Bugzilla::ModPerl->apache_config($cgi_path);
+    $server->add_config( [ grep { length $_ } split( "\n", $conf ) ] );
+
+
+    # Pre-load all extensions
+    Bugzilla::Extension->load_all();
+
+    Bugzilla->preload_features();
+
+    # Force instantiation of template so Bugzilla::Template::PreloadProvider can do its magic.
+    Bugzilla->template;
+
+    Bugzilla::ModPerl::BlockIP->set_memcached( Bugzilla::Memcached->_new()->{memcached} );
```

Comment:
it's weird to use a private constructor and then reach into the object to get the memcached var.  i see that's how the old code worked, but as you're touching how and when this object is created it would be nice (ie. optional) to fix this now.

i'd prefer to see a public method added to `Bugzilla::Memcache` which returns the internal `{memcached}` object, and give `Bugzilla->memcached` the ability to return the non-metric'd version upon request.

Reply:
I just made the constructor public. The metrics code can probably be removed in a subsequent PR, I don't think it would be easy to get running again given changes to elasticsearch and so on.